editor/src/components/canvas/dom-walker.ts

```diff
@@ -428,6 +436,135 @@ function runSelectiveDomWalker(
   }
 }
 
+export interface BackfillDomMetadataResult {
+  updatedMetadata: ElementInstanceMetadataMap
+  reconstructedMetadata: ElementInstanceMetadataMap
+}
+
+export function backfillDomMetadata(
+  metadata: ElementInstanceMetadataMap,
+): BackfillDomMetadataResult {
+  // Find the paths missing from the metadata, which are ancestors of paths in the metadata which are not present in the metadata.
+  const missingPaths = new Set<string>()
+  let parentsAndChildrenToRebuild: { [parent: string]: Array<string> } = {}
+  for (const elementMetadata of Object.values(metadata)) {
+    function addPathToParentsAndChildren(pathToAdd: ElementPath): void {
+      const parentPath = EP.parentPath(pathToAdd)
+
+      const parentPathString = EP.toString(parentPath)
+      // Build up parentsAndChildrenToRebuild.
+      let pathsToFillChildren: Array<string>
+      if (parentPathString in parentsAndChildrenToRebuild) {
+        pathsToFillChildren = parentsAndChildrenToRebuild[parentPathString]
+      } else {
+        pathsToFillChildren = []
+        parentsAndChildrenToRebuild[parentPathString] = pathsToFillChildren
+      }
+      pathsToFillChildren.push(EP.toString(pathToAdd))
+    }
+    function fillPath(pathToFill: ElementPath): void {
+      addPathToParentsAndChildren(pathToFill)
+      if (!EP.isEmptyPath(pathToFill)) {
+        const pathStringToFill = EP.toString(pathToFill)
+        if (!missingPaths.has(pathStringToFill) && !(pathStringToFill in metadata)) {
+          missingPaths.add(pathStringToFill)
+          const parentPath = EP.parentPath(pathToFill)
+          if (parentPath != null) {
+            fillPath(parentPath)
+          }
+        }
+      }
+    }
```

Comment:
These functions should live outside of the `for` loop

Reply:
Fixed in aeae4fd8f.